### ISSUE TITLE
[Feature Request] Autosave

### ISSUE THREAD

**Issue by l0wl3vel** [UberWriter/uberwriter#176]
**Is your feature request related to a problem? Please describe.**
An autosave feature would be really nice, since shutting down the computer deletes all progress after the last manual save.

**Describe the solution you'd like**
An option in the settings to do autosaves every X minutes.

**Comment by somas95** [comment]
I've been thinking about this and how to implement it. 

### Pros: 
It prevents data loss 

### Cons:
You may save something you don't want to save. Without proper version control XXXX

Possible implementations:

## Hot saving
Like Notepad++ 
It saves automatically on exit, even if the user didn't save the file originally. Reopening the application retrieves the state of it when it was closed

## Backup files
The file saving is managed by the user, but the application also saves a .md~ one every x minutes. In the event of a crash, power loss, etc, the application can recover the "lost" file from the backup one

## Autosaving
Literally, autosaving the file each x minutes. The user still can manually save nonetheless. Some kind of feedback must be done (when was my file last saved?) and also a way to reverse to a previous manual save

**Comment by wolfv** [comment]
I think hot saving is cool! sublime text also does something similar. I don't like "backup files" that much where these file are cluttering the folder though (as what happens with gedit). Maybe it could be possible to store the hot save files somewhere else (e.g. `$HOME/.local/usr/uberwriter/<HASH_OF_FILE>.md`) where `HASH_OF_FILE` is some hash computed from the contents of the original file + location on the disk? That way, the autosave would also automatically invalidate if the file is opened from another editor and modified there...

Ok that were just some random thoughts on this topic :) maybe too much.

**Comment by somas95** [comment]
Then hot saving it is :)

But the implementation should be future proof. I.e, if in the future we add multi instance support, do we only save the state of the last closed window? 

Another option is to save on exit but to provide a "recent edited" window when launching uberwriter the next time (like Builder does) 

I really like the idea of keeping track of the hash of the open files to know if they have been edited by other applications (Gedit does this) but I don't think a "hash file" is necessary for that unless we want to track changes between uberwriter launches (which I think we shouldn't) 

So, to recap, I'd 

- Save ever (?) secods/minutes 
- Save when losing focus 
- Hashing when losing focus 
- Checking hash when recovering focus 
- Save when closing the application 

On launching we have several options: 

- To show a "recents screen" (see Gitg, Builder, etc)
- To open the last file edited (what do we do if in the future we support multiple instances?) 
- To open a blank file, letting the user to select a recent one with the current recents menu

**Comment by Frivolous2** [comment]
I think there should be a togglable option "Open last saved file on startup". The default option would be to always open a blank file. If the optionis toggled, the first instance opens the last file, and each new instance opens a blank file. 

Saving every 5 seconds sounds ideal to me. It worries me a bit from an SSD weardown perspective, but I'm not sure saving that small an amount of text would inflict any serious long-term damage. If the file is not changed, a.k.a. the writer is just sitting there staring at the screen, no saving is needed.

**Comment by somas95** [comment]
You can open the last saved file from the recents menu, in any case the ability to have persistency is interesting and we're considering it. 

Saving each 5s is not a problem for modern SSDs, but is not desirable... You could save unwanted changes or even lose a lot of work (typical example: cat walking on keyboard)

Some sort of autosaving will be eventually implemented, but the implementation will be a little more complex than just saving each c seconds/minutes

**Comment by Frivolous2** [comment]
Cat walking on keyboard is easily reversed by Ctrl + Z.
Recently saved files do not appear to work on flatpak build/Fedora 31.
I think a not bad idea could be "save after 2s of inactivity", so when the user stops typing and thinks, it gets saved. But I personally don't think autosave every 5s is bad unless both an accident happens AND the program crashes/computer crashes, because the user can Ctrl + Z otherwise. 
In that case, what if UberWriter saves Ctrl + Z history per document in its own data storage? Should be communicated to users who think they can transport the .md file and keep the history, but would otherwise solve a lot of data corruption/accident issues, because the user can simply undo them.
